src/tempo.py

```diff
@@ -345,6 +338,52 @@ def generate_config(self, receivers: Sequence[ReceiverProtocol]) -> dict:
 
         return config
 
+    def _build_storage_config(self, s3_config: Optional[dict] = None):
+        if s3_config:
+            return {
+                "trace": {
+                    "backend": "s3",
+                    "s3": {
+                        "bucket": s3_config["bucket"],
+                        "access_key": s3_config["access-key"],
+                        # remove scheme to avoid "Endpoint url cannot have fully qualified paths." on Tempo startup
+                        "endpoint": re.sub(
+                            rf"^{urlparse(s3_config['endpoint']).scheme}://",
+                            "",
+                            s3_config["endpoint"],
+                        ),
+                        "secret_key": s3_config["secret-key"],
+                        "insecure": (
+                            False if s3_config["endpoint"].startswith("https://") else True
+                        ),
+                    },
+                    "wal": {
+                        # where to store the wal locally
+                        "path": self.wal_path
+                    },
+                    "pool": {
+                        # number of traces per index record
+                        "max_workers": 400,
+                        "queue_depth": 20000,
+                    },
+                }
+            }
+        return {
+            "trace": {
+                "backend": "local",
+                "local": {"path": "/traces"},
+                "wal": {
+                    # where to store the wal locally
+                    "path": self.wal_path
+                },
+                "pool": {
+                    # number of traces per index record
+                    "max_workers": 400,
+                    "queue_depth": 20000,
+                },
+            }
+        }
+
```

Comment:
```suggestion
    def _build_storage_config(self, s3_config: Optional[dict] = None):
        trace_config = {
            "wal": {"path": self.wal_path},  # where to store the wal locally
            "pool": {
                "max_workers": 400,  # number of traces per index record
                "queue_depth": 20000,
            },
        }

        if s3_config:
            trace_config.update(
                {
                    "backend": "s3",
                    "s3": {
                        "bucket": s3_config["bucket"],
                        "access_key": s3_config["access-key"],
                        # remove scheme to avoid "Endpoint url cannot have fully qualified paths." on Tempo startup
                        "endpoint": urlparse(s3_config["endpoint"]).netloc,
                        "secret_key": s3_config["secret-key"],
                        "insecure": not s3_config["endpoint"].startswith("https://"),
                    },
                }
            )
        else:
            trace_config.update(
                {
                    "backend": "local",
                    "local": {"path": "/traces"},
                }
            )

        return {"trace": trace_config}
```

Reply:
refactored